Add ability to evolve multiple Pokémon

DevilJamJar opened the pull request:
Adds the ability to evolve multiple Pokémon at once using `commands.Greedy`.

The number of Pokémon is capped at 30, and anything less than 20 Pokémon will be represented as fields, while anything between 20 and 30 will be in the embed description. Thumbnail sprites will only be used when singular Pokémon are evolved.

Resolves #113

pikaninja commented:
having the *, args would make the Greedy attempt to convert say "1 2 3 4 5" into one pokemon, instead of having each one go into a separate pokemon, you cannot have the * there.

DevilJamJar commented:
Thank you, I’ve removed the asterisk

pikaninja commented:
adding on to this, there's almost no reason to use a greedy here as the last argument, having `*args: converter`, achieves the same goal, though I'm not sure which one is easier to read.

DevilJamJar commented:
I thought about that too, and the only reason I chose to use greedy is since it is used in other commands to the same effect, like battle add, so for the sake of uniformity I did it here too

pikaninja commented:
i see.

oliver-ni commented on cogs/pokemon.py:
```diff
@@ -949,42 +949,57 @@ async def get_page(source, menu, pidx):
 
-        if pokemon is None:
+        if not all(pokemon is not None for pokemon in args):
             return await ctx.send("Couldn't find that pokémon!")
```
Can we specify which Pokémon here?

DevilJamJar replied:
Not entirely sure how to add this seeing as `pokemon` is None here, and the `args: commands.Greedy` doesn't store the initial unconverted argument

oliver-ni commented on cogs/pokemon.py:
```diff
@@ -949,42 +949,57 @@ async def get_page(source, menu, pidx):
         guild = await self.bot.mongo.fetch_guild(ctx.guild)
 
-        if (evo := pokemon.get_next_evolution(guild.is_day)) is None:
-            return await ctx.send("That pokémon can't be evolved!")
```
Same here probably

DevilJamJar replied:
Unless I'm reading this wrong, I believe this is on line 973?

Wiper-R reviewed (changes requested):
You must check `len(args) < 1` as well as use `pokemon:n` instead of `str(pokemon.species) + f"{pokemon.nickname}"`.

oliver-ni reviewed (changes requested):
Small nitpick

pikaninja commented:
another thing, you check if len(pokemon) < 30 inside each iteration of the loop, it may be better to just add an if statement at the top above the for loop that returns and sends if it over 30? So it's slightly clearer imo.

DevilJamJar commented:
I've moved the over 30 if statement out of the loop and changed `pokemon` to `args`
> use `pokemon:n` instead of `str(pokemon.species) + f"{pokemon.nickname}"`

@Wiper-R Could you please add a line number and expand on this please, I'm not sure I fully understand, sorry

Wiper-R commented:
> I've moved the over 30 if statement out of the loop and changed `pokemon` to `args`
> 
> > use `pokemon:n` instead of `str(pokemon.species) + f"{pokemon.nickname}"`
> 
> @Wiper-R Could you please add a line number and expand on this please, I'm not sure I fully understand, sorry

Refer to line 970 in `pokemon.py`,
https://github.com/oliver-ni/poketwo/blob/master/cogs/mongo.py#L72-L96

DevilJamJar commented:
Interesting, I never knew about that, but before this, the same method I'm using was used, so I assume that they didn't want extra clutter in the Pokémon names, and so they used this instead? Correct me if I'm wrong, but it was copied from the original so I assume it's preferred for this?

oliver-ni commented:
> Interesting, I never knew about that, but before this, the same method I'm using was used, so I assume that they didn't want extra clutter in the Pokémon names, and so they used this instead? Correct me if I'm wrong, but it was copied from the original so I assume it's preferred for this?

That code predates the `__format__` implementation that i wrote to make things nicer, and I never noticed/changed it. It's probably better to use the format string now.

DevilJamJar commented:
Thanks for the help, I believe I've done that correctly, in terms of the requested changes, should I change the use of commands.Greedy and handle the conversion inside of the command in order to keep the unconverted args for error throwing, or is there another way to do that?

DevilJamJar commented:
I realised it's probably better to use format() here instead of str.format so I changed that.

oliver-ni commented:
Thank you!
